Sources/UDP/UDPSocket.swift

```diff
@@ -0,0 +1,114 @@
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import Core
+import CLibvenice
+import POSIX
+
+public enum UDPError: Error {
+    case didSendDataWithError(error: SystemError, remaining: Data)
+    case didReceiveDataWithError(error: SystemError, received: Data)
+}
+
+extension UDPError: CustomStringConvertible {
+    public var description: String {
+        switch self {
+        case .didSendDataWithError(let error, _): return "\(error)"
+        case .didReceiveDataWithError(let error, _): return "\(error)"
+        }
+    }
+}
+
+public final class UDPSocket {
+    private var socket: udpsock?
+    public private(set) var closed = false
+
+    public var port: Int {
+        return Int(udpport(socket))
+    }
+
+    internal init(socket: udpsock) {
+        self.socket = socket
+    }
+
+    public convenience init(ip: IP) throws {
+        guard let socket = udplisten(ip.address) else {
+            throw SystemError.socketTypeNotSupported
+        }
+        try ensureLastOperationSucceeded()
+        self.init(socket: socket)
+    }
+
+    deinit {
+        if let socket = socket, !closed {
+            udpclose(socket)
+        }
+    }
+
+    public func send(_ data: Data, ip: IP, timingOut deadline: Double = .never) throws {
+        try ensureStreamIsOpen()
+
+        data.withUnsafeBytes { (ptr: UnsafePointer<Data>) -> Void in
+            udpsend(socket, ip.address, ptr, data.count)
+        }
+
+        try ensureLastOperationSucceeded()
+    }
+
+    public func receive(upTo byteCount: Int, timingOut deadline: Double = .never) throws -> (Data, IP) {
+        try ensureStreamIsOpen()
+
+        var address = ipaddr()
+        // var data = Data.buffer(with: byteCount)
+        var data = Data(count: byteCount)
+
+        let received = data.withUnsafeMutableBytes { (ptr: UnsafeMutablePointer<Data>) -> Int in
+            udprecv(socket, &address, ptr, data.count, deadline.int64milliseconds)
+        }
+
+        let receivedData = Data(data.prefix(received))
+
+        do {
+            try ensureLastOperationSucceeded()
+        } catch let error as SystemError where received > 0 {
+            throw UDPError.didReceiveDataWithError(error: error, received: receivedData)
+        }
+
+        let ip = IP(address: address)
+        return (receivedData, ip)
+    }
+
+    public func close() throws {
```

Comment:
IIRC `close()` on `TCP` and `File` doesn't throw. Maybe we should do the same here?